enterprise/cmd/executor/internal/worker/handler.go

```diff
@@ -31,6 +32,31 @@ type handler struct {
 }
 
 var _ workerutil.Handler = &handler{}
+var _ workerutil.WithPreDequeue = &handler{}
+
+// PreDequeue determines if the number of VMs with the current instance's VM Prefix is less than
+// the maximum number of concurrent handlers. If so, then a new job can be dequeued. Otherwise,
+// we have an orphaned VM somewhere on the host that will be cleaned up by the background janitor
+// process - refuse to dequeue a job for now so that we do not over-commit on VMs and cause issues
+// with keeping our heartbeats due to machine load. We'll continue to check this condition on the
+// polling interval
+func (h *handler) PreDequeue(ctx context.Context) (dequeueable bool, extraDequeueArguments interface{}, err error) {
+	if !h.options.FirecrackerOptions.Enabled {
+		return true, nil, nil
+	}
+
+	currentVMs, err := ignite.CurrentlyRunningVMs(context.Background(), h.options.VMPrefix)
+	if err != nil {
+		return false, nil, err
+	}
+
+	if len(currentVMs) < h.options.WorkerOptions.NumHandlers {
+		return true, nil, nil
+	}
+
+	log15.Warn("Orphaned VMs detected - refusing to dequeue a new job", "numCurrentVMs", len(currentVMs), "numHandlers", h.options.WorkerOptions.NumHandlers)
```

Comment:
```suggestion
	log15.Warn("Orphaned VMs detected - refusing to dequeue a new job until it's cleaned up", "numCurrentVMs", len(currentVMs), "numHandlers", h.options.WorkerOptions.NumHandlers)
```

maybe makes it a bit easier to understand this should be a transient error